### PR TITLE
[#617] Extended 'ContentTrait' with node access grants rebuild steps

### LINKED ISSUE #617: Extend 'ContentTrait' with node access grants rebuild steps.
# Extend `ContentTrait` with node access grants rebuild steps

## Summary

Add two node access grants rebuild steps to the existing
`src/Drupal/ContentTrait.php` — one for a specific node by type + title,
one for all nodes in the system.

## Motivation

Any site running a node access module (core `node.module` access grants,
`group`, `entity_access_by_field`, `workbench_access`, `permissions_by_term`)
needs to rebuild grants after test fixtures are created. Without a
rebuild, `SELECT … FROM node_access` returns an empty row set for test
nodes and all access checks fall through to the fallback "no access"
row, breaking scenarios that assert a user can or cannot see a node.

Currently tests resort to `drush php-eval 'node_access_rebuild();'` via a
custom step. No reusable step exists.

These steps belong in `ContentTrait` — not a dedicated trait — because:

1. The by-title variant requires a node lookup by type and exact title,
   which is exactly the pattern `ContentTrait` already uses for
   `contentVisitViewWithTitle`, `contentVisitEditPageWithTitle`,
   `contentAssertNotExistsWithTitle`, etc.
2. Access grants are a node-specific concern — the scope of the trait.
3. Consumers already using `ContentTrait` get the functionality with no
   extra `use` statement.
4. Avoids trait proliferation for two short methods.

## Proposed steps

```gherkin
When I rebuild the access grants for the "article" content with the title "My article"
When I rebuild the access grants for all content
```

### Step definitions

| Step | Method |
|---|---|
| `When I rebuild the access grants for the :content_type content with the title :title` | `contentRebuildAccessGrantsByTitle` |
| `When I rebuild the access grants for all content` | `contentRebuildAccessGrantsAll` |

## Implementation notes

- Add two methods to existing `src/Drupal/ContentTrait.php`.
- `contentRebuildAccessGrantsByTitle(string $content_type, string $title)`:
  1. Load the node by content type + exact title match using the same
     entity query pattern used by `contentVisitViewWithTitle`.
  2. Throw if no node found.
  3. Call `\Drupal::entityTypeManager()->getAccessControlHandler('node')
     ->acquireGrants($node)`.
- `contentRebuildAccessGrantsAll()`: call `node_access_rebuild(FALSE)`
  (non-batched).
- Do **not** support "last node created" — drevops style prefers explicit
  title lookup, matching the rest of `ContentTrait`.
- Steps are `When` because rebuilding grants is an action, not a state
  precondition (`Given` conventions reserve `exists`/`have` wording for
  preconditions — see `CONTRIBUTING.md`).
- Placeholder name is `:content_type` (not `:type`) per `CONTRIBUTING.md`
  — matches every other step in `ContentTrait`.
- Method names: `contentRebuildAccessGrantsByTitle`,
  `contentRebuildAccessGrantsAll` — consistent with existing
  `content*` naming convention.

## Scope boundaries

- No per-module grant debugging steps (e.g. "grant rid X for node Y
  should exist"). Those are too implementation-specific.
- No tag-based auto-rebuild-before-scenario — keep it explicit.
- Non-node entity access rebuilds (block_content, media, taxonomy) are
  out of scope — those entity types don't have grant tables in the same
  way `node_access` does.

## Acceptance criteria

- [ ] Two new steps in `src/Drupal/ContentTrait.php`.
- [ ] Steps use `#[When(...)]` tuple annotations.
- [ ] Placeholder name is `:content_type` (not `:type`).
- [ ] Method names `contentRebuildAccessGrantsByTitle` /
      `contentRebuildAccessGrantsAll` — consistent with the existing
      `content*` naming convention in the trait.
- [ ] Throws a clear exception when no node matches the title.
- [ ] Feature tests added to the existing
      `tests/behat/features/drupal_content.feature`.
- [ ] Uses a simple test scenario: enable a node access module, create a
      node, rebuild grants, verify access checks return the expected
      result.
- [ ] `@trait:ContentTrait` negative tests.
- [ ] `ahoy update-docs` run.
- [ ] `ahoy lint` passes.

### PR DESCRIPTION
**Pull request by AlexSkrypnyk**
## Summary

Extends `ContentTrait` with two `When` steps for rebuilding Drupal node access grants from within Behat scenarios. These are useful for tests that enable a node access module (core `node.module` grants, `group`, `entity_access_by_field`, `workbench_access`, `permissions_by_term`) and need to populate grant rows before asserting visibility.

- `When I rebuild the access grants for the :content_type content with the title :title` — loads the node by type + exact title and calls `acquireGrants()` via the node access control handler.
- `When I rebuild the access grants for all content` — calls `node_access_rebuild(FALSE)` (non-batched).

Fixes #617

## Acceptance checklist

- [x] Two new steps in `src/Drupal/ContentTrait.php`.
- [x] Steps use `#[When(...)]` tuple annotations.
- [x] Placeholder name is `:content_type` (not `:type`).
- [x] Method names `contentRebuildAccessGrantsByTitle` / `contentRebuildAccessGrantsAll` — consistent with the existing `content*` naming convention.
- [x] Throws a clear exception when no node matches the title.
- [x] Feature tests added to `tests/behat/features/drupal_content.feature`.
- [x] `@trait:ContentTrait` negative test for missing content.
- [ ] `ahoy update-docs` run (deferred — handled in follow-up polish phase).
- [ ] `ahoy lint` passes (deferred — handled in follow-up polish phase).

## Merge order

This PR should be merged **after** #615, which also touches `src/Drupal/ContentTrait.php` and `tests/behat/features/drupal_content.feature`. A rebase onto the post-#615 `main` will be performed in the follow-up polish phase.

### Overview
Extends `ContentTrait` with two new Behat step definitions to rebuild Drupal node access grants from test scenarios. This enables populating `node_access` grant rows after enabling node access modules, allowing visibility assertions to function correctly.

### Changes

**src/Drupal/ContentTrait.php**
- Added `contentRebuildAccessGrantsByTitle(string $content_type, string $title): void`
  - Step: `When I rebuild the access grants for the :content_type content with the title :title`
  - Loads a node by content type and exact title match, then acquires grants via the node access control handler
  - Throws `RuntimeException` if no matching node is found
- Added `contentRebuildAccessGrantsAll(): void`
  - Step: `When I rebuild the access grants for all content`
  - Triggers non-batched rebuild of node access grants for all nodes via `node_access_rebuild(FALSE)`
- Introduced `NodeAccessControlHandlerInterface` dependency
- Both methods use `#[When(...)]` tuple annotations per coding standards

**tests/behat/features/drupal_content.feature**
- Added positive `@api` scenario: verifies rebuild by title step correctly acquires grants and content is visitable
- Added positive `@api` scenario: verifies rebuild all content step works and content is visitable
- Added negative `@trait:Drupal\ContentTrait` scenario: validates exception thrown when attempting to rebuild grants for non-existent content

**STEPS.md**
- Documented both new step definitions with examples

### Compliance
✓ All step definitions comply with CONTRIBUTING.md guidelines:
  - Use tuple format `#[When(...)]` instead of regular expressions
  - Use descriptive placeholder names (`:content_type`)
  - Method names begin with trait name (`content*`)
  - When steps follow action format ("When I rebuild...")
  - Use proper exception handling for edge cases

### Testing
✓ Feature tests included for both positive and negative scenarios
✓ Follows existing ContentTrait conventions for node loading via entity query

### Outstanding
- Documentation update (`ahoy update-docs`) and linting (`ahoy lint-docs`) deferred to follow-up polish PR
- Merge contingent on prior merge of #615; rebase planned in follow-up polish

### CODE REVIEW

File: src/Drupal/ContentTrait.php [hunk 1]
```diff
@@ -231,6 +232,59 @@ public function contentChangeModerationStateWithTitle(string $content_type, stri
...
+
+    // @codeCoverageIgnoreStart
+    if (!$node instanceof NodeInterface) {
+      throw new \RuntimeException(sprintf('Unable to find "%s" content with title "%s".', $content_type, $title));
+    }
+    // @codeCoverageIgnoreEnd
+    $handler = \Drupal::entityTypeManager()->getAccessControlHandler('node');
+    assert($handler instanceof NodeAccessControlHandlerInterface);
+    $handler->acquireGrants($node);
+  }
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply src/Drupal/ContentTrait.php:271]: Fixed in e33005b. Captured the grants returned by acquireGrants() and persisted them via node.grant_storage->write().
